### ISSUE TITLE
Why are some bower_components not in vendor.js and instead publicly available?

### ISSUE THREAD

**Issue by rreitmann** [jhipster/generator-jhipster#2284]
I just saw that /bower_components/angular-i18n/angular-locale_en.js is exposed to the web and not minified and packaged in vendor.js (see here https://metadatamanagement.cfapps.io/bower_components/angular-i18n/angular-locale_en.js)

Is this on purpose? I assumed all bower_components are minified and packad in vendor.js and are therefore not even part of the packaged war (prod profile).

**Comment by jdubois** [comment]
I didn't think they could be packaged: so yes it is on purpose, but if I'm wrong let's package them!

**Comment by rreitmann** [comment]
Hmm, I am new to the grunt universe :-) and can't see where it builds vendor.js. So what is currently packaged into vendor.js?

**Comment by andidev** [comment]
There may be a reason for this. The locale seems to be loaded dynamically see 
https://github.com/jhipster/generator-jhipster/blob/master/app/templates/src/main/webapp/scripts/app/_app.js#L118

**Comment by rreitmann** [comment]
Ok, that is a reason for anguler-locale_{locale}.js. But if I unzip the war which is deployed, I see all bower_components although they are packaged in vendor.js...

**Comment by andidev** [comment]
Not sure why they are there but my guesses arethat it was to much work to remove all other files without removing the angular-locale files.

**Comment by rreitmann** [comment]
I see but IMHO all bower_components should be within vendor.js and removed from the war...

**Comment by andidev** [comment]
Yes that sound right. The bower_components folder is huge and polutes the war with 23,7 MB (war is totally 60MB for a new project).

Do you feel like making a PR? I think that it is maven war packaging plugin that should be configured to ignore some files but this is just a wild guess.

Maybe this stackoverflow answer will do it?
http://stackoverflow.com/questions/3750189/maven2-excluding-directory-from-war

Though we need to make sure that the `bower_components/angular-i18n/` folder is kept.

**Comment by andidev** [comment]
The [packagingExcludes](https://maven.apache.org/plugins/maven-war-plugin/war-mojo.html#packagingExcludes) configuration seems to take a regex which could be written to exclude all but `bower_components/angular-i18n/` folder

**Comment by atomfrede** [comment]
I've used the angular gulp generator recently. I think they are replacing specific parts when generating the production html (will have a look tomorrow).  

Furthermore, why don't we try to put bower_components into target (or build for gradle)? And also create the dist for the frontend in Java output directories?

**Comment by andidev** [comment]
@atomfrede 
the production index.html does not contain references to bower_components. Thats not the issue. It correctly only references vendor.js. But maybe I missunderstod something that you ment.

bower_components into target is maybe not a good idea cause then it will be removed each `mvn clean` and then it needs to be downloaded again with bower (but maybe it is cached?). Though I think the reccomendation is to have bower_components version controlled. Think this has been discussed before.

**Comment by gmarziou** [comment]
There is a [main-bower-files](https://github.com/ck86/main-bower-files) that iterates on bower dependencies and builds a list of mandatory files a specified in the main property of each bower.json file. This list can then be used for any processing: copy, minification, ... 

I think this is how it is done in [generator-m-ionic](https://github.com/mwaylabs/generator-m-ionic)'s gulpfile that I use for ionic apps. It works also for grunt.

So if we would put bower_components outside the webapp folder, we could use this task to copy only required files into war file.

Alternatively, we could bower_components in src/main/webapp  but use this task to build target/webapp for production only.

**Comment by deepu105** [comment]
Guys if you deploy the war with dev profile then js and css are served from
bower components. Also unless we include source maps for minified files its
a good idea to serve as it is in dev profile for debugging.

Im finding it very difficult to debug in prod profile. I opened an issue
for this

**Comment by atomfrede** [comment]
@andidev Yes, you're right, we have decided to have bower_components in VCS (so forget my point with putting it into target). I think the problem is, that ``/bower_components/angular-i18n/angular-locale_en.js`` is loaded even in production, which sould not be the case (and bower_components is not required in prod war file). @rreitmann correct me if I got you wrong

**Comment by rreitmann** [comment]
@atomfrede you perfectly got my point. I think the packagingExcludes approach mentioned by @andidev is the right way to go but only for prod profile. In addition we need to take care that ``` /bower_components/angular-i18n/angular-locale_{locale}.js``` is packaged withing vendor.js and loaded here https://github.com/jhipster/generator-jhipster/blob/master/app/templates/src/main/webapp/scripts/app/_app.js#L118

**Comment by andidev** [comment]
@rreitmann though I'm not sure `/bower_components/angular-i18n/angular-locale_{locale}.js` can be packaged in `vendor.js`. I think there is a reason why it is loaded dynamically. If you look at the source code of [angular-locale_en.js](https://cdnjs.cloudflare.com/ajax/libs/angular-i18n/1.4.6/angular-locale_en.js) it looks like it can only be one language placed in $locale.
That means that if we would place all `angular-locale_xx.js` files in vendor.js they would overwrite each other and it would not be possible to change the $locale dynamically on language changes.

So in other words the `angular-locale_xx.js` files can never be placed in `vendor.js` since they need to be loaded dynamically from the app.

**Comment by jdubois** [comment]
Yes for the `bower_components` in prod that's on purpose: some companies want to have the same WAR file in integration testing and in prod, so that makes sense to have a big WAR file with both the dev and the prod builds.
I know it's a bit stupid, and that it fatten the WAR -> maybe this could be an option at build time.

**Comment by jdubois** [comment]
Yes I remember having already worked on this and @andidev is right, I don't think those can be put in the dist folder.

I'm closing this as we have a good reason of doing it that way: of course I'd love to have this improved, so if anybody wants to pick this up and try to make it better, please do.
